MacOS mdns (Bonjour) backend

robcarruthers opened the pull request:
I have made some progress on Issue #39 and would greatly appreciate your guidance and feedback on refining this further. Although my C code might not be perfect, the foundation of the implementation is now functional. Any feedback or suggestions for improvement would be immensely valuable.

### Points of Concern:
- Ensuring that the implementation aligns with the Avahi code, particularly in regards to functions and naming conventions.
- Effective memory management.
- Completion of tasks listed below.

### Progress Based on #39 Suggestions:
- [x] Enabled meson to search for the Bonjour mDNS dependency in the case of macOS.
- [x] Introduced a HAVE_BONJOUR, similar to HAVE_AVAHI.
- [x] Allowed for the calling of the bonjour_discover() function.
- [x] Renamed the example main() to bonjour_discover().
- [x] Add a comprehensive list of service types relevant for LXI. There are 5 types.
- [x] In the example code's resolve_callback(), call the liblxi service callback with the following information: address, id, service, and port. The callback is passed via the lxi_info_t structure as a parameter to bonjour_discover().
- [x] Incorporate the timeout parameter from bonjour_discover() into the timeval tv structure.


I have tested this on macOS 13, with an rPi setup with Avahi to respond to '_lxi._tcp' and '_vxi-11._tcp' broadcasts.

``` shell
❯ clang -I/opt/homebrew/include test/search.c -o search -L/opt/homebrew/lib -llxi

❯ ./search
Searching for LXI devices - please wait...

Browse callback
Service name: dummy-scope
Regtype: _lxi._tcp.
Domain: local.

Broadcasting on interface Found service
Resolve callback
Full name: dummy-scope._lxi._tcp.local.
Host target: dummy-scope.local.
Port: 12345

fish: Job 1, './search' terminated by signal SIGBUS (Misaligned address error)
```

lundmar commented:
Great. Looks good, we just need to sort out the remaining tasks and a few details.

lundmar commented on src/bonjour.c:
```diff
@@ -0,0 +1,102 @@
+#include "bonjour.h"
+#include "lxi.h"
+
+char *services[] = {
+    "_lxi._tcp",
+    "_vxi-11._tcp",
+};
```
Here we just need to add "_scpi-raw._tcp", "_scpi-telnet._tcp", "_hislip._tcp".

lundmar replied:
According to https://developer.apple.com/documentation/dnssd/1804742-dnsservicebrowse we can put these in one string with comma separation so it will search for the services by calling DNSServiceBrowse() once. No array needed.

robcarruthers replied:
I've added the extra service types as a comma separated string, but the DNSServiceBrowse function returns a error code -65540 (from [this code](https://github.com/pieceofsummer/TouchRemote/blob/master/foo_touchremote/TouchRemote.Bonjour/DnsServiceErrorType.cs), it would indicate a bad param)

So I added a loop as per the Avahi code, but this seems to only broadcast the first string in the array because this is the only response back from the device. ie. all other types are ignored or not broadcast. Is this likely because the info pointer is not copied for each iteration?

``` c
void create_service_browser(char *reg_type, lxi_info_t *info) {
    DNSServiceRef service;
    DNSServiceErrorType error;

    error = DNSServiceBrowse(&service, 0, 0, reg_type, "local.", browse_callback, info);
    if(error != kDNSServiceErr_NoError) {
        fprintf(stderr, "DNSServiceBrowse() failed: %d\n", error);
        return;
    }

    error = DNSServiceProcessResult(service);
    if(error != kDNSServiceErr_NoError) {
        fprintf(stderr, "DNSServiceProcessResult() failed: %d\n", error);
        return;
    }

    DNSServiceRefDeallocate(service);
}

int bonjour_discover(lxi_info_t *info, int timeout) {
    DNSServiceRef service;
    DNSServiceErrorType error;
    int i;
    int service_count = sizeof(services) / sizeof(services[0]);

    // Loop over services
    for (i = 0; i < service_count; i++) {
        // create service browser
        create_service_browser(services[i], info);
    }
        return LXI_OK;
}
```

lundmar commented:
Oh, I just noticed that you have scrapped the timeout handling code which was included in the original example main function. We need that back. Also, the new code is iterating over the 5 services calling DNSServiceBrowse() for each service. In the original code it was only called once to search for e.g. 2 services. I expect it should be possible to expand that string with the next services and still only cann DNSServiceBrowse() once?

robcarruthers commented:
I've got too many things not working here.

Could you detail how the `lxi_info_t` struct is used and how it is expected to flow through from `bonjour_discover` to the callbacks ie. when does the service callback get assigned to it and when is it called?. I think my implementation has this screwed up because I've isolated the SIGBUS (Misaligned address error) above to this line in `resolve_callback`
``` c
info->service(fullname, hosttarget, "Resolved service", ntohs(port));
```

lundmar commented:
Oh, I understand now. Please don't look too much to the Avahi code. The Avahi code will only confuse and these are two different implementations so we should not expect them to be structured exactly the same.

I think it is best to go back to the original example that you tested. It did work searching for 2 services right using a single comma separated string? We get that working the implementation will be better.

robcarruthers commented:
The original code gives an error: Browse failed with error code -65540. Its hard to find docs for this code but, as [above](https://github.com/lxi-tools/liblxi/pull/40#discussion_r1195886449), it seems to indicate a bad param.

Once I split them into an array it worked. but only the first service broadcast was responded to.

I'm happy to revert the code as required, each time I get more familiar with it.

lundmar commented:
The lxi_info contains 3 callbacks, 2 called during normal discovery, and 1, the service() callback, is called during mDNS discovery. In context of your Bonjour implementation it should be called from your resolve_callback() which you are already doing:

https://github.com/lxi-tools/liblxi/pull/40/files#diff-fbee693ed57176f747acc91a48d7b8df6e3a0c19f7290f58d835f6101bf9b74dR28-R29

To me it looks like you are passing the point to the lxi_info structure correctly via the "context" parameters.

In the end the lxi tool will take the information you pass with the callback and present it like this example:

```
lundmar@wopr:~$ lxi discover -m
Searching for LXI devices - please wait...

  Found "Rohde&Schwarz power supply NGM202-101403" on address 192.168.0.107
    lxi service on port 80
  Found "Rohde&Schwarz power supply NGM202-101403" on address 192.168.0.107
    scpi-raw service on port 5025

Found 2 services
```

In your resolve_callback() this information includes your variables fullname, hosttarget, port... we are only missing the service type name, which it looks like we maybe have to pass from the browse_callback() to the resolve_callback().

lundmar commented:
Oh, so the original example didn't work.

Ok, I've reread the doc for DNSServiceBrowse() and I think I read it too quickly. It is not a comma separated list. It used dot separation but that is only to device one service. So you are doing it correctly iterating through the 5 services using a string array.

lundmar commented on src/bonjour.c:
```diff
@@ -0,0 +1,102 @@
+        }
+
+        DNSServiceRefDeallocate(service);
+        return LXI_OK;
```
This line is the reason your code is only searching for the first service in your services array. You don't want to return from the function here. Simply move the return statement down to the end of your function and you should be fine.

lundmar commented:
Oh, I just realized that you are testing with the search.c test application and not the lxi tool itself. To make that work you need to add the service() callback function. As you can see, it is missing from https://github.com/lxi-tools/liblxi/blob/63939e01184c7ff4ad647fd5e024ec801ee788c9/test/search.c

lundmar commented:
It may just be better if you test with the lxi tool as I demonstrated in the example output.

robcarruthers commented:
> It may just be better or easier if you test with the lxi tool as I demonstrated in the example output.

I agree, however I haven't manage to get the cli running in macOS just yet.
I'll look at pulling out the discover function from lxi tools and run that as a test file instead of the search example.

lundmar commented:
I see. First things first. All you need to do is add this to your search.c:
```c
static void service(const char *address, const char *id, const char *service, int port)
{
    printf("  Found \"%s\" on address %s\n    %s service on port %u\n", id, address, service, port);
}
```
That will fix your crash issue.

robcarruthers commented:
I'm making slow progress with an adjusted version of your example code and derived from other examples and the [dns_sd.h file](https://opensource.apple.com/source/mDNSResponder/mDNSResponder-544/mDNSShared/dns_sd.h.auto.html). I should have started there first... I can be a bit stupid sometimes!

The following output needs refinement, but runs without errors and adjusts to the number of services. So the service strings can be an array, declared somewhere.

``` shell
Searching for LXI devices - please wait...

  Found "dummy-scope.local." on address dummy-scope\032LXI._lxi._tcp.local.
    Resolved service service on port 12345

  Found "dummy-scope.local." on address dummy-scope\032VXI-11._vxi-11._tcp.local.
    Resolved service service on port 111

  Found "dummy-scope.local." on address dummy-scope\032SCPI._scpi-raw._tcp.local.
    Resolved service service on port 5025

  Found "dummy-scope.local." on address dummy-scope\032SCPI\032Telnet._scpi-telnet._tcp.local.
    Resolved service service on port 5026

  Found "dummy-scope.local." on address dummy-scope\032HiSLIP._hislip._tcp.local.
    Resolved service service on port 4880
```

I'm working getting the required data available in the info callbacks and then formatting the output as per the cli (with the search.c file). With regards to the info.service params, the resolve_callback code doesn't return an IP address. But we could do a lookup with hosttarget, what's the preferred method to do this? is it always an FQDN (ie. dummy-scope.local.) or can it be a hostname?

Next I'll then add in the timeout functionality. This is likely to generate more questions...

I've got a busy few days, so may not have this ironed out until next week. The code is still similar to the your sample but different enough, that it may not make much sense adding to this PR. How best to handle this?

Also, please don't think by diverging your example I'm disregarding you help, I really value the time and direction you are providing, thank you. I'm also super keen to get this working as I think it's a great addition to the overall product.

lundmar commented:
> I'm making slow progress with an adjusted version of your example code and derived from other examples and the [dns_sd.h file](https://opensource.apple.com/source/mDNSResponder/mDNSResponder-544/mDNSShared/dns_sd.h.auto.html). I should have started there first... I can be a bit stupid sometimes!
> 

No worries! You do it your way :)

C is like riding a bike.. sometimes it takes a little practice and maybe detours to get riding again :)

> The following output needs refinement, but runs without errors and adjusts to the number of services. So the service strings can be an array, declared somewhere.
> 
> ```shell
> Searching for LXI devices - please wait...
> 
>   Found "dummy-scope.local." on address dummy-scope\032LXI._lxi._tcp.local.
>     Resolved service service on port 12345
> 
>   Found "dummy-scope.local." on address dummy-scope\032VXI-11._vxi-11._tcp.local.
>     Resolved service service on port 111
> 
>   Found "dummy-scope.local." on address dummy-scope\032SCPI._scpi-raw._tcp.local.
>     Resolved service service on port 5025
> 
>   Found "dummy-scope.local." on address dummy-scope\032SCPI\032Telnet._scpi-telnet._tcp.local.
>     Resolved service service on port 5026
> 
>   Found "dummy-scope.local." on address dummy-scope\032HiSLIP._hislip._tcp.local.
>     Resolved service service on port 4880
> ```
> 

Great, you got it going.

I wonder where that '032' stuff comes from. That seems a bit out of place.

> I'm working getting the required data available in the info callbacks and then formatting the output as per the cli (with the search.c file). With regards to the info.service params, the resolve_callback code doesn't return an IP address. But we could do a lookup with hosttarget, what's the preferred method to do this? is it always an FQDN (ie. dummy-scope.local.) or can it be a hostname?

You can use the getaddrinfo() to retrieve the host IP from a host name address string. It should understand any FQDN or partial FQDN you throw at it. You can find many examples using that. Though, I think "dummy-scope.local." should be "dummy-scope.local" but maybe it understands regardless.

> 
> Next I'll then add in the timeout functionality. This is likely to generate more questions...
> 

Of course, the select() mechanism is not the easiest of constructs to understand.

> I've got a busy few days, so may not have this ironed out until next week. The code is still similar to the your sample but different enough, that it may not make much sense adding to this PR. How best to handle this?
> 

No hurry, we are not slave to any schedule in these open source matters :)

Regarding git, there are various ways to go about that. It is fine to keep the PR.

Whatever changes you made, you can simply keep pushing to this PR branch even though the commit comments may need cleanup. Then in the end, I should be able to squash the PR into one commit which I think is fine for this feature.

You can also just update/overwrite your latest commit by using the 'git --amend' command and then make a 'git push --force' to overwrite.

Either way, we can clean up the commits once you have things working.

> Also, please don't think by diverging your example I'm disregarding you help, I really value the time and direction you are providing, thank you. I'm also super keen to get this working as I think it's a great addition to the overall product.

Oh don't you worry at all about that. I consider it all part of your journey to understand the code. I simply appreciate your effort contributing to this project so we can finally have mac OS support for them mac-ies ;)

robcarruthers commented:
Great that helps a lot.

I need some help creating the info callback data. The problem I have is that only some data is available in the `DNSServiceBrowseReply browse_callback` function and I need it in the `DNSServiceResolveReply resolve_callback` to give to the `info->service` callback.

For example I need `regtype` and `servicename`, as these correspond to `service` and `id` params (or are they args in C).
I know this is simple, but how do I store these strings and access them later, ensuring they are still valid?

[DNSServiceBrowseReply](https://developer.apple.com/documentation/dnssd/dnsservicebrowsereply?language=objc)
[DNSServiceResolveReply](https://developer.apple.com/documentation/dnssd/dnsserviceresolvereply?language=objc)

I need to get the interface name using the interfaceIndex variable too, but I think I can find a function for that.

lundmar commented:
Yes, so you basically want to pass on data from your browse_callback to the resolve_callback. The way to do that is, instead of passing a pointer to the info structure, you want to pass a pointer to a newly allocated structure that embeds the pointer to the info structure and copies of whatever else information you want to pass. For example, first define the new structure:

``` c
typdef struct
{
   lxi_info_t *info;
   char *servicename;
   char *regtype;
} browse_data_t;
```

The, in your browse callback you prepare the browse_data structure to be passed via DNSServiceResolve():
```c
browse_data_t *browse_data = malloc(sizeof(browse_data_t));
browse_data->info = info;
browse_data->servicename = strdup(servicename);
browse_data->regtype = strdup(regtype);
```

In your resolve_callback you do something like this:
```c
browse_data_t = (browse_data_t *) context;

// Use the data

free(browse_data->regtype);
free(browse_data->servicename);
free(browse_data);
```

robcarruthers commented:
I think this is pretty close. Let me know how best to test for complete functionality. It works with my dummy scope setup and local printers. I haven't tested on Linux yet, but I didn't really touch that code.

The search example needs some formatting but I think it's displaying the correct device values now.

The timeout addition in the `browse_lxi_services` function is GPT code, so just cast your eye over it to ensure it's correct. I've tested it works, however I'm reluctant to trust it 100%.

I've refactored the service strings and labels into a struct and stored them in an array, to make it easier to loop over. I've added this to the lxi.h file so the avahi code could be modified to use this also and save duplication.

I also thought of adding a discover_type param to the search.c example eg. mdns, vxi11 (default)
```
./search mdns
```

``` shell
❯ ./search
Searching for LXI devices - please wait...

Broadcasting on interface en0

  Found "dummy-scope LXI" on address 192.168.10.109
    lxi service on port 12345

  Found "dummy-scope VXI-11" on address 192.168.10.109
    vxi-11 service on port 111

  Found "dummy-scope SCPI" on address 192.168.10.109
    scpi-raw service on port 5025

  Found "dummy-scope SCPI Telnet" on address 192.168.10.109
    scpi-telnet service on port 5026

  Found "dummy-scope HiSLIP" on address 192.168.10.109
    hislip service on port 4880
```

Please let me know what you think and feedback any changes/improvements.
What other examples or docs need updating?
Is there anything on the LXI Tools side that need updating?

I've created a [lxi_rb](https://github.com/robcarruthers/lxi_rb/tree/master) rubygem that wraps the liblxi library, works on the Mac and Linux too. I'll get it updated for bonjour discovery over the next few days.

lundmar commented:
Great 👍🏻 

Looks good to me except one thing... Please don't destroy the original search test code. Rather create a new search-mdns.c. These only serve as simple liblxi code examples.

lundmar commented:
I don't think there is anything we need to update in lxi-tools. If you compile lxi-tools for mac the lxi commandline tool should just work.

lundmar commented:
Finally, it would be great if you could squash all your commits into one feature patch 👍🏻

Commit:
feat(bonjour): Add mdns support for macOS

This code is designed to discover LXI (LAN eXtensions for Instrumentation) devices using the Bonjour service (dns_sd.h) discovery protocol. The flow of the implementation is as follows:

1. bonjour_discover(lxi_info_t *info, int timeout_ms) is the entry point of the code. It takes an lxi_info_t struct to store the discovered information and a timeout in milliseconds. It calls the browse_lxi_services() function with these parameters. The lxi_info_t is setup with callbacks to initially to recieve the discovered instrument data. See test folder for examples.

2. lxi_info_t struct is defined in lxi.h.
browse_lxi_services(lxi_info_t *info, int timeout_ms) iterates through lxi_services (defined in lxi.h) and calls the Bonjour DNSServiceBrowse() function for each service. It then processes the results and sets a timeout to stop waiting for results after a specified duration. It provides the browse_callback() function as a callback.

3. browse_callback() is called for each discovered service. It extracts relevant information about the service and then creates a browse_data_t struct to aggregate service data. It calls DNSServiceResolve() to resolve the service's hostname, and provides resolve_callback() as a callback.

4. resolve_callback() is called once the service hostname is resolved. It calls the resolve_ip_address() function to convert the hostname into an IP address. It also extracts more information about the service and then calls the info->service() callback function with the discovered service's information. Finally, it frees the memory allocated for the browse_data_t struct.

5. resolve_ip_address() takes a hostname and port, and converts them into an IP address using the getaddrinfo() function. The resulting IP address is stored in the provided buffer.

Service types and labels are declared in the lxi_services array in  src/lxc.c

robcarruthers commented:
I'm not sure I done the right thing forcing this push, but I had trouble squashing the commits as requested and retaining my changes. The current push has one commit, but I'm not sure the commit history has come through, anyway the commit has enough detail added about the changes made to cover most queries.

I'll take a look at getting LXI Tools on macOS working now (or at the very least the CLI)  and then document the process, somehow. How would you like this done?

lundmar commented on test/search-mdns.c:
```diff
@@ -0,0 +1,33 @@
+#include <stdio.h>
+#include <lxi.h>
+
+// Example - Search for LXI instruments using VXI-11 discovery
```
VXI-11 -> mDNS :)

lundmar commented on src/bonjour.c:
```diff
@@ -0,0 +1,186 @@
+
+    for (lxi_service_t *s = lxi_services; s->broadcast_type != NULL; s++)
+    {
+        error = DNSServiceBrowse(&service, 0, 0, s->broadcast_type, "local.", browse_callback, info);
```
I think we should just use NULL here instead "local." as the documentation states the following:
```
domain

    If non-NULL, specifies the domain on which to browse for services. Most applications will not specify a domain, instead browsing on the default domain(s).
```

This way it falls back to using whatever is hosts default.

lundmar commented:
Looks like your squash made it just fine 👍🏻 

I did a final read through and found two minor things to correct. If you fix those I will merge.

Commit:
refactor(bonjour): Fix typo & Refine browse default domain
